R/dependencies.R

```diff
@@ -0,0 +1,410 @@
+##' @title Get the dependencies for a given report from the database
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##' @param con A connection to a database
+get_dependencies_db <- function(name, id, upstream, con, list_all = FALSE) {
+  ## now construct the SQL query
+  if (upstream) {
+    filt_qry <- sprintf("depends.report_version='%s'", id)
+  } else {
+    filt_qry <- sprintf("report_version.id='%s'", id)
+  }
+
+  sql_qry <- paste(c("SELECT",
+               "depends.report_version, ",
+               "report_version.report, report_version.id, ",
+               "file_artefact.filename, ", "file_artefact.file_hash",
+               "FROM", "(depends",
+               "INNER JOIN", "file_artefact", "ON",
+               "depends.use=file_artefact.id",
+               "INNER JOIN", "report_version_artefact", "ON",
+               "file_artefact.artefact=report_version_artefact.id",
+               "INNER JOIN", "report_version", "ON",
+               "report_version_artefact.report_version=report_version.id)",
+               "WHERE", filt_qry), collapse = " ")
+
+  db_ret <- DBI::dbGetQuery(con, sql_qry)
+
+  if (nrow(db_ret) == 0) {
+    return(NULL)
+  }
+
+  db_ret$is_latest <- sapply(db_ret$report_version,
+                               is_latest_in_db, con = con)
+  if (!list_all) {
+    db_ret <- db_ret[which(db_ret$is_latest), ]
+  }
+
+  # if we're going uptree
+  if (upstream) {
+    return(unique(db_ret$id))
```

Comment:
please do not use explicit return statements if it is not an early exit (here and elsewhere)